server/src/controllers/PartyController.ts

```diff
@@ -0,0 +1,345 @@
+        res
+          .status(400)
+          .send({ type: "duplicateKey", address: originalAddress });
+        return;
+      }
+      PartyController.sendUnknownError(res, err);
+    }
+  };
+
+  public get = async (req: Request, res: Response) => {
+    const address = req.params.address;
+    await this.getByAddress(address, res, party =>
+      PartyController.mapAndReturnPartyData(address, res, party)
+    );
+  };
+
+  public validatePassword = async (req: Request, res: Response) => {
+    const address = req.params.address;
+    const password = req.body.password;
+
+    if (!password && address) {
+      res.json(false);
+      return;
+    }
+
+    const party = await this.partyCollection.findOne({
+      address: PartyController.trimAndToLower(address)
+    });
+    if (!party) {
+      res.status(404).send({ type: "notFound", address });
+    } else {
+      res.json(party.password === password);
+    }
+  };
+
+  public signupToParty = async (req: Request, res: Response) => {
+    const userAddress = PartyController.trimAndToLower(req.body.userAddress);
+    const grail = await this.grailCollection.findOne({
+      address: userAddress
+    });
+    if (!grail) {
+      res.status(404).send({ type: "notFound", userAddress });
+      return;
+    }
+    const partyAddress = PartyController.trimAndToLower(req.params.address);
+    const result = await this.partyCollection.findOneAndUpdate(
+      {
+        address: partyAddress,
+        userlist: { $nin: [userAddress] },
+        pendingUserlist: { $nin: [userAddress] }
+      },
+      {
+        $addToSet: {
+          pendingUserlist: userAddress
+        }
+      },
+      { returnOriginal: false }
+    );
+    if (result && result.ok && result.value) {
+      res.json({ success: true });
+      return;
+    }
+    // We didnt get a result, so there must already be a user with this address in the specified party
+    res.status(409).send({ type: "conflict", userAddress });
+  };
+
+  public updatePartyUser = async (req: Request, res: Response) => {
+    try {
+      const method = req.body.method;
+      switch (method) {
+        case "accept":
+          this.acceptPartyUser(req, res);
+          break;
+        case "deny":
+          this.denyPartyUser(req, res);
+          break;
+        case "remove":
+          this.removePartyUser(req, res);
+          break;
+        default:
+          res.status(404).send({ type: "methodUnknown", method });
+      }
+    } catch (err) {
+      PartyController.sendUnknownError(res, err);
+    }
+  };
+
+  private acceptPartyUser = async (req: Request, res: Response) => {
+    const password = req.body.password;
+    const address = PartyController.trimAndToLower(req.body.address);
+    const token = req.body.token;
+    const acceptedUser = req.body.user;
+    const result = await this.partyCollection.findOne({
+      address: address,
+      password: password,
+      token: token
```

Comment:
the three methods have really a lot of code duplication, would be nice if you could extract the common logic somehow, however, can also be done later => issue

Reply:
Yes, agreed. This is updated in the latest commit, check it out!